### ISSUE TITLE
[aot autograd][inline inbuilt nn modules] AOT Autograd changes _version different from eager

### ISSUE THREAD

**Issue by anijain2305** [pytorch/pytorch#128198]
### 🐛 Describe the bug

The test case fails when Dynamo inlines the builtin nn modules. It passes for backend "eager" but fails for backend "aot_eager"

~~~

import torch
import copy
import operator


torch._dynamo.config.inline_inbuilt_nn_modules = True

class MyModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.m = torch.nn.BatchNorm1d(100)

    def forward(self, x):
        return self.m(x)

model_for_eager = MyModel()
model_for_compile = copy.deepcopy(model_for_eager)

eager_version_counters = [
    buffer._version for _, buffer in model_for_eager.named_buffers()
]
compile_version_counters = [
    buffer._version for _, buffer in model_for_compile.named_buffers()
]

compiled_f = torch.compile(model_for_compile, backend="aot_eager")

inp_ref = torch.ones(20, 100, requires_grad=True)
inp_test = torch.ones(20, 100, requires_grad=True)

out_ref = model_for_eager(inp_ref.clone())
out_test = compiled_f(inp_test.clone())

eager_version_counters_after = [
    buffer._version for _, buffer in model_for_eager.named_buffers()
]
compile_version_counters_after = [
    buffer._version for _, buffer in model_for_compile.named_buffers()
]

breakpoint()
eager_delta = list(
    map(operator.sub, eager_version_counters_after, eager_version_counters)
)
compile_delta = list(
    map(operator.sub, compile_version_counters_after, compile_version_counters)
)

assert eager_delta == compile_delta
~~~

### Versions

NA

cc @ezyang @bdhirsh @chauhang @zou3519 @msaroufim

**Comment by anijain2305** [comment]
cc @bdhirsh @zou3519

**Comment by zou3519** [comment]
Er, is this the thing where aten::batch_norm has incorrect schema? Or did we fix that?

**Comment by ezyang** [comment]
We don't guarantee that you will get the same vc bump count, just that it is bumped

**Comment by anijain2305** [comment]
Ah, I see .. so I should just change the test, right?

**Comment by zou3519** [comment]
@ezyang the problem is that some of the version counters weren't bumped in the eager case, while they were bumped in the compile case

**Comment by zou3519** [comment]
@anijain2305 and I dug into this some more. Here's what's going on:
- eager-mode batch_norm does not bump the version counters of the buffers. This is a long-standing issue (https://github.com/pytorch/pytorch/issues/13402, https://github.com/pytorch/pytorch/issues/120622). What's the status on this? (cc @andrewor14, @janeyx99)
- torch.compile(backend=aot_eager) decomposes batch_norm into some pointwise operations and a copy_ back into the original buffer. This does bump the version counter (correctly).
- torch.compile(backend=inductor) is interesting. It seems like we did work in https://github.com/pytorch/pytorch/pull/111046 to ensure that the copy_ stays in the inductor graph and becomes a part of a fused kernel. Does this mean that we go from bumping the version counter to not bumping the version counter anymore? (cc @bdhirsh, @tugsbayasgalan)

Anyways my read of the correct behavior is: since we intend for batch_norm to bump the version counters of the buffers in eager-mode PyTorch, it is OK for torch.compile to bump the version counters while we fixup eager-mode PyTorch to match.

**Comment by bdhirsh** [comment]
I think this behavior actually makes sense (https://github.com/pytorch/pytorch/pull/131665 also fixes the problem for inference, but doesn't affect the repro above since it generates a training graph).

When I run the above repro with the **inductor** backend, I get:
```
(Pdb) p eager_version_counters
[1, 1, 1]
(Pdb) p compile_version_counters
[1, 1, 1]
(Pdb) p eager_version_counters_after
[1, 1, 2]
(Pdb) p compile_version_counters_after
[2, 2, 2]
```

The compile(backend="inductor") version is correct: all 3 of batch_norm's buffers were mutated and had their VC's bumped (`running_mean`, `running_var`, `num_batches_tracked`). Eager is wrong - the first two VC's haven't been bumped, because the eager mode impl doesn't bump the version counters correctly (tracked here: https://github.com/pytorch/pytorch/issues/120622)

Running with `backend=aot_eager`, I instead get:
```
(Pdb) p compile_version_counters_after
[3, 3, 3]
```

Why did we bump the VC twice?

(1) we bump the VC of each buffer once at runtime here, where are effectively calling "mark-dirty" on each input that we saw was mutated ([link](https://github.com/pytorch/pytorch/pull/131665/files#diff-6ff2ea4c2706e1aa3306f23a6caa435f458bbc4d0cb2fdd32c46de5d24ee30d2R305))

(2) when using `backend="aot_eager"` the actual code we invoke at runtime is an FX graph with `aten.copy_(buffer, updated_buffer)` in it. Since this is an aten op, it will **also** bump the VC of `buffer`, so we each buffer gets its VC bumped twice.

Compare this to inductor: inductor generates a raw triton kernel for the input mutations. Since the triton kernel is not an ATen op, it does not go through the dispatcher and bump the VC of the inputs that it mutates.

One way to make the VC more consistent would be for AOTAutograd to detect whether we are running with aot_eager and not bump the VC. But in general that seems pretty fragile. I'm hoping we can just decide that it is not important to bump the VC the exact same number of times between compile and eager (and the only invariant we need to maintain is "VC on every mutated buffer goes up by some number in torch.compile")

**Comment by zou3519** [comment]
> I'm hoping we can just decide that it is not important to bump the VC the exact same number of times between compile and eager

We don't need to bump the VC the exact same number of times between compile and eager, it just needs to be bumped some amount

**Comment by StrongerXi** [comment]
> We don't need to bump the VC the exact same number of times between compile and eager, it just needs to be bumped some amount

Doing old issue scrubbing; the original repro still fails on main d0f08dc3ebc, but sounds like that's due to an overly restrictive check?

@anijain2305 maybe we can close this issue and reenable the relevant test (it already passes for me)?
https://github.com/pytorch/pytorch/blob/f3304571fcf84c9779c3861001e787e5dff53988/test/inductor/test_torchinductor.py#L4823-L4824
